optuna/integration/cma.py

```diff
@@ -0,0 +1,438 @@
+
+        complete_trials = []
+        for trial in trials:
+            if trial.distributions != self._search_space:
+                continue
+            if trial.state != TrialState.COMPLETE:
+                continue
+            if not self._is_compatible(trial):
+                continue
+            complete_trials.append(trial)
+
+        popsize = self._es.popsize
+        generation = len(complete_trials) // popsize
+        for i in range(generation):
+            xs = []
+            ys = []
+            for t in complete_trials[i * popsize:(i + 1) * popsize]:
+                x = [
+                    self._to_cma_params(self._search_space, name, t.params[name])
+                    for name in self._param_names
+                ]
+                xs.append(x)
+                ys.append(t.value)
+            if study_direction == StudyDirection.MAXIMIZE:
+                ys = [-1 * y if y is not None else y for y in ys]
+            self._es.ask()
+            self._es.tell(xs, ys)
+        return generation * popsize
+
+    def ask(self, trials, n_told):
+        # type: (List[FrozenTrial], int) -> Dict[str, Any]
+
+        individual_index = self._n_target_trials(trials) - n_told
+        popsize = self._es.popsize
+
+        # individual_index may exceed the population size when users execute multiple trials in
+        # parallel. Note that trial may suggest the same parameters when multiple samplers invoke
+        # this method simultaneously.
+        while individual_index >= popsize:
+            individual_index -= popsize
+            self._es.ask()
+        cma_params = self._es.ask()[individual_index]
+
+        ret_val = {}
+        for param_name, value in zip(self._param_names, cma_params):
+            ret_val[param_name] = self._to_optuna_params(self._search_space, param_name, value)
+        return ret_val
+
+    def _is_compatible(self, trial):
+        # type: (FrozenTrial) -> bool
+
+        # Thanks to `product_search_space()` function, in sequential optimization,
+        # the parameters of complete trials are always compatible with the search space.
+        #
+        # However, in distributed optimization, incompatible trials may complete on a worker
+        # just after a product search space is calculated on another worker.
+
+        for name, distribution in self._search_space.items():
+            if name not in trial.params:
+                return False
+
+            distributions.check_distribution_compatibility(distribution, trial.distributions[name])
+            param_value = trial.params[name]
+            param_internal_value = distribution.to_internal_repr(param_value)
+            if not distribution._contains(param_internal_value):
+                return False
+
+        return True
+
+    def _n_target_trials(self, trials):
+        # type: (List[FrozenTrial]) -> int
+
+        cnt = 0
+        for trial in trials:
+            if trial.distributions != self._search_space:
```

Comment:
This logic seems different from [L321](https://github.com/pfnet/optuna/pull/447/files#diff-45cb45846c0a7c8fcb996f62ffbba36cR321).
For ensuring consistency, how about defining (for example) `collect_target_trials(study)` method and passing the result trials to `tell` and `ask` methods?

Reply:
Thank you for your suggestion. I extracted the common procedure to [`_collect_target_trials` method](https://github.com/pfnet/optuna/pull/447/files#diff-45cb45846c0a7c8fcb996f62ffbba36cR394-R402). Please note that I added some filters to support parallel execution of trials.